### PR TITLE
[Experimental] Diffusion LoRA DPO training

### PR DESCRIPTION
**Pull request by sayakpaul**
Based on the amazing work done by @bram-w here for Diffusion DPO: https://github.com/bram-w/trl/blob/dpo/. The ultimate goal is to have it in `trl`. 

Uses this toy dataset for training validation: https://huggingface.co/datasets/kashif/pickascore. 

Training command:

```bash
CUDA_VISIBLE_DEVICES=0 accelerate launch train_diffusion_dpo.py \
  --pretrained_model_name_or_path=runwayml/stable-diffusion-v1-5  \
  --output_dir="diffusion-dpo" \
  --mixed_precision="fp16" \
  --dataset_name=kashif/pickascore \
  --resolution=512 \
  --train_batch_size=16 \
  --gradient_accumulation_steps=2 \
  --gradient_checkpointing \
  --use_8bit_adam \
  --rank=8 \
  --learning_rate=1e-5 \
  --report_to="wandb" \
  --lr_scheduler="constant" \
  --lr_warmup_steps=0 \
  --max_train_steps=10000 \
  --checkpointing_steps=2000 \
  --run_validation --validation_steps=200 \
  --seed="0" \
  --report_to="wandb" \
  --push_to_hub
```

WandB (currently training): https://wandb.ai/sayakpaul/diffusion-dpo-lora/runs/v055hlxa?workspace=user-sayakpaul

Cc: @kashif @lvwerra

### CODE REVIEW

**Comment by sayakpaul** [comment]
For 10k steps and just 500 samples and with an LR of 1e-4, it's probably now overfitting beyond 3k steps. But it seems to be clearly having its effect! @kashif any comments?

**Comment by bram-w** [comment]
This looks great, thanks so much for putting it together! Checked out the loss logs and they look reasonable; nice to see that stability for the main loss. The relative MSEs indicate that probably the KL-beta would be higher to be optimal, but the run's in a funky low-data regime and code-wise it looks super healthy.

Full-finetuning/SDXL would both be great to have as well but for most research use cases (especially academic) this is the perfect starting point.

**Comment by sayakpaul** [comment]
Will copy over what I said on Slack for housekeeping:

Thanks for your comments and good luck! I think we can put together a nice SDXL script too. We need to add micro-conditioning support and we should be done. Good thing is we have got plenty of references for doing that nicely.

@kashif, would you be able to give that a try?

**Comment by bram-w** [comment]
Awesome, sounds great!

**Comment by sayakpaul** [comment]
I added support for SDXL too 🤗

Training command:

```bash
CUDA_VISIBLE_DEVICES=3 accelerate launch train_diffusion_dpo_sdxl.py \
  --pretrained_model_name_or_path=stabilityai/stable-diffusion-xl-base-1.0  \
  --pretrained_vae_model_name_or_path=madebyollin/sdxl-vae-fp16-fix \
  --output_dir="diffusion-sdxl-dpo" \
  --mixed_precision="fp16" \
  --dataset_name=kashif/pickascore \
  --train_batch_size=8 \
  --gradient_accumulation_steps=2 \
  --gradient_checkpointing \
  --use_8bit_adam \
  --rank=8 \
  --learning_rate=1e-5 \
  --report_to="wandb" \
  --lr_scheduler="constant" \
  --lr_warmup_steps=0 \
  --max_train_steps=2000 \
  --checkpointing_steps=500 \
  --run_validation --validation_steps=50 \
  --seed="0" \
  --report_to="wandb" \
  --push_to_hub
```

WandB page: https://wandb.ai/sayakpaul/diffusion-dpo-lora-sdxl/runs/95dnjqmp.

Let's see how this one goes!

@bram-w if you have time, I'd appreciate some review on the script. Major point of review would be on how I am handling data preprocessing. During training, SDXL uses something called micro-conditioning which includes the original size, crop coordinates, etc. I have taken the simplest reasonable bet and resized the image corresponding to "jpg_1" to the resolution of the image belonging to "jpg_0".

Let's see how the runs turns out to be 🤞

**Comment by sayakpaul** [comment]
Hmm, seems like there's numerical issues with the VAE. This is why all the generated images are coming out to be black here: https://wandb.ai/sayakpaul/diffusion-dpo-lora-sdxl/runs/95dnjqmp. Will debug deeper and make necessary changes.

**Comment by bram-w** [comment]
@sayakpaul 

For the HFlip: https://github.com/huggingface/diffusers/pull/6468

Cheers!

**Comment by Youngon** [comment]
> Hmm, seems like there's numerical issues with the VAE. This is why all the generated images are coming out to be black here: https://wandb.ai/sayakpaul/diffusion-dpo-lora-sdxl/runs/95dnjqmp. Will debug deeper and make necessary changes.
> 
> (**_Update_**): the bug is fixed. New log: https://wandb.ai/sayakpaul/diffusion-dpo-lora-sdxl/table?workspace=user-sayakpaul

I have trained my own lora model , how to use this model to do inference?

**Comment by sayakpaul** [comment]
You can see an example here: 
https://github.com/huggingface/diffusers/blob/a517f665a45bc25a24268aad9997c027769220a7/examples/research_projects/diffusion_dpo/train_diffusion_dpo_sdxl.py#L93
